{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2606540b",
   "metadata": {},
   "source": [
    "# Single Cell sequencing (scRNA-seq)\n",
    "\n",
    "This notebook consist of:\n",
    "- Data QC (at cell and genes level)\n",
    "- Visualisation (embedding)\n",
    "- Machine learning \n",
    "\n",
    "We utilized the PBMC (peripheral blood mononuclear cells)datasets and mostly utilized Scanpy (equivalent to Seurat in R)\n",
    "\n",
    "# Environment setup \n",
    "The scRNA usually required multiple files which make it better to run it inside a conda/mamba environment to prevent any damage to the local computer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35b9df78",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Environment setup; succesful loading here should mean all dependencies correctly installed #\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "\n",
    "from scipy import stats\n",
    "from scipy.cluster.hierarchy import dendrogram\n",
    "\n",
    "import seaborn as sbn\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from gprofiler import GProfiler\n",
    "\n",
    "from natsort import natsorted\n",
    "from sklearn.cluster import AgglomerativeClustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c90e3d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Suppress warnings for tidy representation of the notebook   \n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "class bcolors:\n",
    "    HEADER = '\\033[95m'\n",
    "    OKBLUE = '\\033[94m'\n",
    "    OKGREEN = '\\033[92m'\n",
    "    WARNING = '\\033[93m'\n",
    "    FAIL = '\\033[91m'\n",
    "    ENDC = '\\033[0m'\n",
    "    BOLD = '\\033[1m'\n",
    "    UNDERLINE = '\\033[4m'\n",
    "    \n",
    "def print_msg(msg_lst: list, sep='\\t', color='\\033[91m'):\n",
    "    msg = \"\"\n",
    "    for i in msg_lst:\n",
    "        msg += str(i) + sep\n",
    "\n",
    "    print(color + \"-\" * 80 + bcolors.ENDC)\n",
    "    print(color + msg.center(80, ' ') + bcolors.ENDC)\n",
    "    print(color + '-' * 80 + bcolors.ENDC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "041062a4",
   "metadata": {},
   "source": [
    "## Loading the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f33e3e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#############################\n",
    "#   Loading the Data   # \n",
    "#############################\n",
    "pbmc = sc.read_10x_mtx(\n",
    "    'data/filtered_gene_bc_matrices/hg19/',  # the directory with the `.mtx` file\n",
    "    var_names='gene_symbols',                # use gene symbols for the variable names (variables-axis index)\n",
    "    cache=True)                              # write a cache file for faster subsequent reading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8714a0f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Understand the data structure\n",
    "print(pbmc.X.getnnz(axis = 1)) #2700 x 32738\n",
    "print(pbmc.var)               #32738 genes\n",
    "print(pbmc.obs)               #2700 cells\n",
    "print(pbmc.uns)               #unstructured annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29af3f22",
   "metadata": {},
   "source": [
    "# QC\n",
    "## QC at cell level \n",
    "There is usually some variations between cells due to poor separation of oil droplet (the scRNA-seq usually used droplet-based method) and sequencing error. \n",
    "\n",
    "### QC metrics using ribosomal cell \n",
    "High expression of ribosomal genes indicates failure to capture other expressed genes within the cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f467cf6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create an observation metric info to add the total reads per cell\n",
    "obs_metrics = pd.DataFrame(index=pbmc.obs_names) ## Get the identifiers\n",
    "\n",
    "print(\"Total Rb genes \", sum(pbmc.var_names.str.startswith(('RPL', 'RPS')))) # Number of ribosomal genes\n",
    "\n",
    "# Calculate QC metrics for ribosomal genes \n",
    "pbmc.var['rb'] = pbmc.var_names.str.startswith(('RPL', 'RPS'))\n",
    "sc.pp.calculate_qc_metrics(\n",
    "    pbmc,\n",
    "    qc_vars=['rb'],\n",
    "    percent_top=None,\n",
    "    log1p=False,\n",
    "    inplace=True\n",
    ")\n",
    "\n",
    "# Calculate the RB cells >50%\n",
    "print(\"total RB cells >50%:\", (pbmc.obs['pct_counts_rb'] > 50).sum())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06e4bdc9",
   "metadata": {},
   "source": [
    "### QC metrics using mitochondrial cells\n",
    "High mitochondrial expression means the cells is dying or almost reach senescences stage, which need to be excluded."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a02726e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Total Mt genes \", sum(pbmc.var_names.str.startswith('MT-'))) # Number of mitochondrial genes\n",
    "\n",
    "pbmc.var[\"MT\"] = pbmc.var_names.str.startswith('MT')\n",
    "\n",
    "sc.pp.calculate_qc_metrics(\n",
    "    pbmc,\n",
    "    qc_vars=['MT'],\n",
    "    percent_top=None,\n",
    "    log1p=False,\n",
    "    inplace=True\n",
    ")\n",
    "\n",
    "print(\"total MT cells >5%:\", (pbmc.obs['pct_counts_MT'] > 5).sum())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50c1363a",
   "metadata": {},
   "source": [
    "## QC at genes level\n",
    "Some low amplification might be missed during sequencing which result to lot of zeros in the cell.\n",
    "\n",
    "See how many genes are expressed in no cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8687becc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get number of non-zero values in each row (gene) and add to annData object. \n",
    "pbmc.var['cells_per_gene'] = pbmc.X.getnnz(axis = 0) \n",
    "genes_in_no_cells = sum(pbmc.var['cells_per_gene'] == 0)\n",
    "print('There are', genes_in_no_cells, 'genes expressed in no cells.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23cb4d71",
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc_copy = pbmc.copy()\n",
    "\n",
    "# Normalize each cell to a total of 1e4 reads\n",
    "sc.pp.normalize_total(pbmc_copy, target_sum=1e4)\n",
    "\n",
    "# Scale it the data to unit variance and zero mean\n",
    "sc.pp.scale(pbmc_copy)\n",
    "\n",
    "# Log transform the data\n",
    "sc.pp.log1p(pbmc_copy) #NOTE this adds a +1 pseudocount to prevent -infinity for log(0)\n",
    "\n",
    "# Run PCA to reduce the dimensionality of the data\n",
    "sc.tl.pca(pbmc_copy, svd_solver='arpack')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ae3784d",
   "metadata": {},
   "source": [
    "# Highly variable genes\n",
    "This is important where they can be our marker genes to distinguish our cells (i.e T cell is CD4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed60b40c",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.highly_variable_genes(pbmc_copy, min_mean=0.0125, max_mean=3, min_disp=0.5)\n",
    "\n",
    "print(\"Number of highly variable genes selected: \", sum(pbmc_copy.var['highly_variable'].values))\n",
    "\n",
    "sc.pl.highly_variable_genes(pbmc_copy)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34e0296b",
   "metadata": {},
   "source": [
    "# Visualisation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e83dd2bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using PCA to see the variance explained by each principal component\n",
    "\n",
    "sc.pl.pca(pbmc_copy, color='LYZ')\n",
    "sc.pl.pca(pbmc_copy, color='CD4', components='2,3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6677088c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using UMAP for visualisation\n",
    "sc.pp.neighbors(pbmc_copy, n_neighbors=10, n_pcs=40)\n",
    "sc.tl.umap(pbmc_copy)\n",
    "sc.pl.umap(pbmc, color=['LYZ','CD3E','MS4A1'] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3d5e020",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use t-SNE for visualisation\n",
    "sc.tl.tsne(pbmc_copy, random_state = 2)\n",
    "sc.pl.tsne(pbmc_copy, color=['LYZ','CD3E','MS4A1'] )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03a97714",
   "metadata": {},
   "source": [
    "## Clustering\n",
    "\n",
    "### k-means clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bf90950",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.cluster import KMeans\n",
    "\n",
    "data_pca = pbmc_copy.obsm['X_pca']\n",
    "\n",
    "# kmeans clustering\n",
    "kmeans_clust = KMeans(n_clusters = 3).fit(data_pca) \n",
    "\n",
    "plt.scatter(data_pca[:, 0], data_pca[:, 1], c=kmeans_clust.labels_)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca5b8f2e",
   "metadata": {},
   "source": [
    "### Hierarchial clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a888a060",
   "metadata": {},
   "outputs": [],
   "source": [
    "HC_clust = AgglomerativeClustering(n_clusters=3, metric='euclidean', linkage='ward').fit(data_pca)\n",
    "\n",
    "plt.scatter(data_pca[:, 0], data_pca[:, 1], c=HC_clust.labels_)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc08d9e7",
   "metadata": {},
   "source": [
    "# Machine learning\n",
    "\n",
    "In here, we only try 2 machine learning which are random forest and MLP.\n",
    "\n",
    "We didn't do the hyperparameter tuning for both ML which can be a limitation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da6fec74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Library \n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from sklearn.metrics import classification_report, accuracy_score, f1_score, precision_score, recall_score\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.neural_network import MLPClassifier\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4ac902f",
   "metadata": {},
   "source": [
    "## Additional QC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c36720cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We did additional QC which is to choose top 2000 highly variable genes for analysis\n",
    "sc.pp.highly_variable_genes(pbmc_copy, n_top_genes=2000)\n",
    "\n",
    "train_pbmc = pbmc_copy[:, pbmc_copy.var['highly_variable']]\n",
    "\n",
    "# visualize the highly variable genes\n",
    "sc.tl.rank_genes_groups(train_pbmc, groupby=\"cell-types\", method=\"wilcoxon\")\n",
    "sc.set_figure_params(scanpy=True, dpi=100, fontsize=14)\n",
    "sc.pl.rank_genes_groups(train_pbmc, n_genes=10, sharey=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "423e9ad9",
   "metadata": {},
   "source": [
    "## Encode the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c282071e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pre-processing encoder\n",
    "X = train_pbmc.X\n",
    "\n",
    "# Encode the Y (for easy use in ML models)\n",
    "mapping = {\"CD56+ NK\" : 1, \n",
    "           \"CD19+ B\": 2, \n",
    "           \"CD4+/CD45RA+/CD25- Naive T\" : 3, \n",
    "           \"Dendritic\":4}\n",
    "\n",
    "train_pbmc.obs[\"cell-encode\"] = train_pbmc.obs[\"cell-types\"].map(mapping)\n",
    "\n",
    "print(train_pbmc.obs[[\"cell-types\", \"cell-encode\"]].head())\n",
    "\n",
    "#Y IS \n",
    "Y = train_pbmc.obs[\"cell-encode\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d429e32",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Train test split\n",
    "X_train, X_test, y_train, y_test = train_test_split(\n",
    "    X, Y, test_size=0.30, random_state=42, stratify=Y\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7557f1a3",
   "metadata": {},
   "source": [
    "## Random forest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e1e6019",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Classifier for random forest \n",
    "rf = RandomForestClassifier(random_state = 42, class_weight=\"balanced\")\n",
    "\n",
    "rf.fit(X_train, y_train)\n",
    "y_train_pred = rf.predict(X_test)\n",
    "\n",
    "inv_map = {v: k for k, v in mapping.items()}\n",
    "target_names = [inv_map[i] for i in sorted(inv_map.keys())]\n",
    "\n",
    "print(\"=== RF Results ===\")\n",
    "print(\"Accuracy:\", accuracy_score(y_test, y_train_pred))\n",
    "print(\"Macro F1:\", f1_score(y_test, y_train_pred, average=\"macro\"))\n",
    "print(\"Precision:\", precision_score(y_test, y_train_pred, average=\"macro\"))\n",
    "print(\"Recall:\", recall_score(y_test, y_train_pred, average=\"macro\"))\n",
    "\n",
    "print(classification_report(y_test, y_train_pred, target_names=target_names))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea17620d",
   "metadata": {},
   "source": [
    "## MLP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "202e55bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MLP classifier \n",
    "mlp = MLPClassifier(random_state=42)\n",
    "\n",
    "mlp.fit(X_train, y_train)\n",
    "y_mlp_pred = mlp.predict(X_test)\n",
    "\n",
    "print(\"=== MLP Results ===\")\n",
    "print(\"Accuracy:\", accuracy_score(y_test, y_mlp_pred))\n",
    "print(\"Macro F1:\", f1_score(y_test, y_mlp_pred, average=\"macro\"))\n",
    "\n",
    "print(classification_report(y_test, y_mlp_pred, target_names=target_names))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce06916b",
   "metadata": {},
   "source": [
    "## Test data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00853eb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import the data\n",
    "test_pbmc = sc.read_h5ad(\"Test_dataset.h5ad\")\n",
    "\n",
    "print(test_pbmc.obs[\"cell-types\"].value_counts())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "825abd85",
   "metadata": {},
   "source": [
    "### QC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2720bca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove non-important cell types\n",
    "# Mitochondrial cells\n",
    "test_pbmc.var[\"mt\"] = test_pbmc.var_names.str.startswith('MT')\n",
    "sc.pp.calculate_qc_metrics(test_pbmc, qc_vars=[\"mt\"], inplace=True)\n",
    "test_pbmc = test_pbmc[test_pbmc.obs[\"pct_counts_mt\"] < 5].copy()\n",
    "\n",
    "# Ribosomal cells\n",
    "test_pbmc.var[\"rb\"] = (\n",
    "    test_pbmc.var_names.str.startswith(\"RPL\") |\n",
    "    test_pbmc.var_names.str.startswith(\"RPS\")\n",
    ")\n",
    "sc.pp.calculate_qc_metrics(test_pbmc, qc_vars=[\"rb\"], inplace=True)\n",
    "test_pbmc = test_pbmc[test_pbmc.obs.pct_counts_mt < 10, :].copy()\n",
    "\n",
    "\n",
    "# Normalize\n",
    "sc.pp.normalize_total(test_pbmc, target_sum=1e4)\n",
    "# Log transform \n",
    "sc.pp.log1p(test_pbmc)\n",
    "# Choose top 2000 highly variables genes \n",
    "sc.pp.highly_variable_genes(test_pbmc, n_top_genes=2000, subset=True)\n",
    "# Scale \n",
    "sc.pp.scale(test_pbmc)\n",
    "# PCA\n",
    "sc.tl.pca(test_pbmc, svd_solver='arpack')\n",
    "# Get the X and Y\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "769ad99f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Encoding\n",
    "# X and Y define \n",
    "X_mys_test = test_pbmc.X\n",
    "\n",
    "# Encode the Y \n",
    "mapping_test = {\"NK cell\" : 1, \n",
    "           \"B cell\": 2, \n",
    "           \"CD4+ T cell\" : 3, \n",
    "           \"CD8+ T cell\": 3,\n",
    "           \"Other T\": 3,\n",
    "           \"cDC\":4,\n",
    "           \"pDC\":4,\n",
    "           \"CD16+ Monocyte\":5,\n",
    "           \"CD14+ Monocyte\" :5,\n",
    "           \"Platelet\":6,\n",
    "           \"Plasmablast\":6}\n",
    "\n",
    "inv_test = {1: \"NK cell\",\n",
    "            2: \"B cell\",\n",
    "            3: \"T cell\",\n",
    "            4: \"Dendritic\",\n",
    "            5: \"Monocytes\",\n",
    "            6: \"Others\"}\n",
    "\n",
    "test_pbmc.obs[\"cell-encode\"] = test_pbmc.obs[\"cell-types\"].map(mapping_test)\n",
    "\n",
    "#Y IS \n",
    "Y_mys_encode = test_pbmc.obs[\"cell-encode\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d46aa42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Exclusion of some cell types not in training data\n",
    "to_exclude = [\"Monocytes\", \"Others\"]\n",
    "filtered_pbmc = test_pbmc[~test_pbmc.obs[\"encode-reverse\"].isin(to_exclude)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f89cdfcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Redefine X and Y\n",
    "# X and Y define \n",
    "X_filt_test = filtered_pbmc.X\n",
    "\n",
    "#Y IS \n",
    "Y_filt_encode = filtered_pbmc.obs[\"cell-encode\"]\n",
    "\n",
    "print(filtered_pbmc.obs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b514dbd",
   "metadata": {},
   "source": [
    "## Random forest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa801923",
   "metadata": {},
   "outputs": [],
   "source": [
    "y_filt_pred = rf.predict(X_filt_test)\n",
    "\n",
    "print(\"=== RF Results ===\")\n",
    "print(\"Accuracy:\", accuracy_score(y_filt_pred, Y_filt_encode))\n",
    "print(\"Macro F1:\", f1_score(y_filt_pred, Y_filt_encode, average=\"macro\"))\n",
    "\n",
    "print(classification_report(y_filt_pred, Y_filt_encode, target_names=target_names))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "054f3afe",
   "metadata": {},
   "source": [
    "## MLP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6edfe39d",
   "metadata": {},
   "outputs": [],
   "source": [
    "y_filt_mlp_pred = mlp.predict(X_filt_test)\n",
    "\n",
    "filtered_pbmc.obs[\"mlp-pred\"] = y_filt_mlp_pred\n",
    "\n",
    "print(\"=== MLP Results ===\")\n",
    "print(\"Accuracy:\", accuracy_score(y_filt_mlp_pred, Y_filt_encode))\n",
    "print(\"Macro F1:\", f1_score(y_filt_mlp_pred, Y_filt_encode, average=\"macro\"))\n",
    "print(\"Precision:\", precision_score(y_filt_mlp_pred, Y_filt_encode, average=\"macro\"))\n",
    "print(\"Recall:\", recall_score(y_filt_mlp_pred, Y_filt_encode, average=\"macro\"))\n",
    "\n",
    "print(classification_report(y_filt_mlp_pred, Y_filt_encode, target_names=target_names))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
